src/lfortran/semantics/ast_symboltable_visitor.cpp

```diff
@@ -1102,7 +1102,13 @@ class SymbolTableVisitor : public CommonVisitor<SymbolTableVisitor> {
         for (size_t i=0; i<x.n_contains; i++) {
             bool current_storage_save = default_storage_save;
             default_storage_save = false;
-            visit_program_unit(*x.m_contains[i]);
+            if ( compiler_options.implicit_typing ) {
+                std::map<std::string, ASR::ttype_t*> implicit_dictionary_copy = implicit_dictionary;
+                visit_program_unit(*x.m_contains[i]);
+                implicit_dictionary = implicit_dictionary_copy;
+            } else {
+                visit_program_unit(*x.m_contains[i]);
+            }
```

Comment:
```suggestion
             std::map<std::string, ASR::ttype_t*> implicit_dictionary_copy = implicit_dictionary;
             visit_program_unit(*x.m_contains[i]);
             implicit_dictionary = implicit_dictionary_copy;
```

We can keep it by default, nice catch!